src/setup-elixir.js

```diff
@@ -1,97 +1,356 @@
 const core = require('@actions/core')
 const {exec} = require('@actions/exec')
-const {installElixir, installOTP} = require('./installer')
+const installer = require('./installer')
 const path = require('path')
 const semver = require('semver')
 const https = require('https')
 const {
-  fstat,
   promises: {readFile},
 } = require('fs')
 
-main().catch(err => {
+try {
+  main().catch((err) => {
+    core.setFailed(err.message)
+  })
+} catch (err) {
   core.setFailed(err.message)
-})
+}
 
 async function main() {
-  checkPlatform()
+  const maybePlatformIsOKError = installer.checkPlatform()
+  if (maybePlatformIsOKError !== 'ok') {
+    throw new Error(maybePlatformIsOKError)
+  }
 
   const osVersion = getRunnerOSVersion()
+  const otpBuildsListing = {
+    remote: `https://repo.hex.pm/builds/otp/${osVersion}/builds.txt`,
+  }
   const otpSpec = core.getInput('otp-version', {required: true})
-  const elixirSpec = core.getInput('elixir-version', {required: true})
-  const otpVersion = await getOtpVersion(otpSpec, osVersion)
-  const [elixirVersion, otpMajor] = await getElixirVersion(
-    elixirSpec,
-    otpVersion
-  )
-
-  let installHex = core.getInput('install-hex')
-  installHex = installHex == null ? 'true' : installHex
+  await installOTP(otpSpec, osVersion, otpBuildsListing)
+    .then(async (otpVersion) => {
+      const elixirBuildsListing = {
+        remote: 'https://repo.hex.pm/builds/elixir/builds.txt',
+      }
+      const elixirSpec = core.getInput('elixir-version', {required: false})
+      await maybeInstallElixir(elixirSpec, otpVersion, elixirBuildsListing)
+        .then(async (elixirInstalled) => {
+          if (elixirInstalled === true) {
+            const shouldMixRebar = core.getInput('install-rebar', {
+              required: false,
+            })
```

Comment:
Is the default still to install rebar and hex and if elixir was installed?

Reply:
I don't completely get the question, since I believe there's an extra "and" there, but if I get it right, yes, `mix` should only execute if Elixir is installed properly. Did I assume this wrong and it should be installed whenever?